File: src/aelongestPalindrome/manacherSolution.java
package aelongestPalindrome;
//

public class manacherSolution {
	public String longestPalindrome(String s){
		int len = s.length();
		if(len == 0||len == 1||len == 2){
			return s;
		}
		
		StringBuilder T = new StringBuilder("#");
		T.append(s.charAt(0));
		for(int i = 1; i < len; i++){
			 T.append('#');
			 T.append(s.charAt(i));
		}
		T.append('#');
		
		int[] P = new int[T.length()];
		
		int R = 1;
		int C = 1;
		int max = 0;
		String retStr = "";
		for(int i = 1; i < T.length() -1; i++){
			int mirror = 2*C - R;
			P[i] = Math.min(R-i,P[mirror]);
			while(i-1-P[i] >= 0 && i + 1 + P[i] < T.length() && (T.charAt(i - 1 - P[i]) == T.charAt(i + 1 + P[i]))){
				P[i] ++;
			}
			if(P[i] > max){
				max = P[i];
				retStr = s.substring((i-P[i])/2,(i+P[i]+1)/2);
			}
			if(i+P[i] > R){
				C = i;
				R = i + P[i];
			}
			
		}
		return retStr;
	}
}

File: src/aelongestPalindrome/manacherSolution.py
# http://articles.leetcode.com/2011/11/longest-palindromic-substring-part-ii.html

class Solution(object):
    def longestPalindrome(self, s):
        """
            :type s: str
            :rtype: str
            """
        if(len(s) == 0 or len(s) == 1 or len(s) == 2):
            return s
        T = "#"
        for i in xrange(len(s)):
            T += s[i]
            T += "#"
        P = [0]*len(T)
        R = 1
        C = 1
        retStr = ""
        max = 0
        for i in xrange(1,len(T)-1):
            image = 2*C - i
            P[i] = min(P[image],R-i)
            while i-1-P[i] >= 0 and i+1+P[i] < len(T) and (T[i-1-P[i]] == T[i+1+P[i]]):
                P[i] += 1
            if P[i] > max:
                max = P[i]
                retStr = s[(i-P[i])/2:(i+1+P[i])/2]
            if P[i] > R-i:
                C = i
                    R = i+P[i]
                return retStr